packages/virglrenderer-android/0009-direct-connection-to-x-server.patch.beforehostbuild

```diff
@@ -0,0 +1,305 @@
+ static inline uint32_t get_buf_entry(const uint32_t *buf, uint32_t offset)
+ {
+    return buf[offset];
++++ ./src/vrend_renderer.h
+@@ -108,6 +108,7 @@
+    uint32_t blob_id;
+    struct list_head head;
+    bool is_imported;
++   uint32_t shm_seg, pixmap;
+ };
+
+ #define VIRGL_TEXTURE_NEED_SWIZZLE        (1 << 0)
+@@ -634,5 +635,8 @@
+ int
+ vrend_renderer_resource_d3d11_texture2d(struct pipe_resource *res, void **handle);
+ 
+ 
++struct vrend_context* vrend_context_get_from_virgl(struct virgl_context* ctx);
++int vrend_renderer_attach_shm_fd(struct pipe_resource *res, int fd);
++int vrend_renderer_draw_to_x_server_drawable(struct pipe_resource *res, struct virgl_context* ctx, uint32_t drawable);
+ #endif
++++ ./src/vrend_renderer.c	2024-11-25 19:44:35.525876727 +0200
+@@ -65,6 +65,11 @@
+
+ #include "tgsi/tgsi_text.h"
+
++#include <fcntl.h>
++#include <xcb/xcb.h>
++#include <xcb/shm.h>
++#include <xcb/xcb_errors.h>
++
+ #ifdef HAVE_EPOXY_GLX_H
+ #include <epoxy/glx.h>
+ #endif
+@@ -368,6 +373,13 @@
+    int gl_major_ver;
+    int gl_minor_ver;
+
++   struct {
++      xcb_connection_t* conn;
++      xcb_errors_context_t *err_ctx;
++      xcb_window_t root;
++      xcb_gcontext_t gc;
++   } xcb;
++
+    mtx_t fence_mutex;
+    thrd_t sync_thread;
+    virgl_gl_context sync_context;
+@@ -7579,6 +7591,22 @@
+    /* make sure you have the latest version of libepoxy */
+    gles = epoxy_is_desktop_gl() == 0;
+
++   vrend_state.xcb.conn = getenv("DISPLAY") ? xcb_connect(NULL, NULL) : NULL;
++   if (xcb_connection_has_error(vrend_state.xcb.conn)) {
++      virgl_error("Failed to connect X server:  %d\n", xcb_connection_has_error(vrend_state.xcb.conn));
++      xcb_disconnect(vrend_state.xcb.conn);
++      vrend_state.xcb.conn = NULL;
++   }
++
++   if (vrend_state.xcb.conn) {
++      xcb_errors_context_new(vrend_state.xcb.conn, &vrend_state.xcb.err_ctx);
++      xcb_screen_t *screen = xcb_setup_roots_iterator(xcb_get_setup(vrend_state.xcb.conn)).data;
++      vrend_state.xcb.root = screen->root;
++      vrend_state.xcb.gc = xcb_generate_id(vrend_state.xcb.conn);
++      xcb_create_gc(vrend_state.xcb.conn, vrend_state.xcb.gc, screen->root, XCB_GC_FOREGROUND | XCB_GC_GRAPHICS_EXPOSURES, (const uint32_t[]) {screen->black_pixel, 0});
++      xcb_flush(vrend_state.xcb.conn);
++   }
++
+    vrend_state.gl_major_ver = gl_ver / 10;
+    vrend_state.gl_minor_ver = gl_ver % 10;
+
+@@ -8846,6 +8874,15 @@
+    if (res->d3d_tex2d)
+       res->d3d_tex2d->lpVtbl->Release(res->d3d_tex2d);
+ #endif
++
++   if (vrend_state.xcb.conn) {
++      if (res->pixmap)
++         xcb_free_pixmap(vrend_state.xcb.conn, res->pixmap);
++      if (res->shm_seg)
++         xcb_shm_detach(vrend_state.xcb.conn, res->shm_seg);
++      xcb_flush(vrend_state.xcb.conn);
++   }
++
+    free(res);
+ }
+
+@@ -12696,6 +12733,9 @@
+    if (has_feature(feat_texture_mirror_clamp_to_edge))
+       caps->v2.capability_bits_v2 |= VIRGL_CAP_V2_MIRROR_CLAMP_TO_EDGE;
+
++   if (vrend_state.xcb.conn)
```

Comment:
Better make it an option so that user can disable it.

Reply:
User can use `unset DISPLAY` before starting `virglrenderer-android` so it will not connect to X server and will not advertise its ability to connect X server. I think I will add an option for virglrenderer-android.